packages/controllers/src/cronjob/CronjobController.ts

```diff
@@ -0,0 +1,342 @@
+import {
+  BaseControllerV2 as BaseController,
+  RestrictedControllerMessenger,
+  HasPermission,
+  GetPermissions,
+} from '@metamask/controllers';
+import {
+  HandlerType,
+  SnapId,
+  TruncatedSnap,
+  CronjobSpecification,
+  flatten,
+  parseCronExpression,
+} from '@metamask/snap-utils';
+import { Duration, inMilliseconds } from '@metamask/utils';
+import {
+  GetAllSnaps,
+  getRunnableSnaps,
+  GetSnap,
+  HandleSnapRequest,
+  SnapAdded,
+  SnapBlocked,
+  SnapEndowments,
+  SnapInstalled,
+  SnapRemoved,
+  SnapTerminated,
+  SnapUnblocked,
+  SnapUpdated,
+} from '..';
+import { getCronjobCaveatJobs } from '../snaps/endowments/cronjob';
+import { Timer } from '../snaps/Timer';
+
+export type CronjobControllerActions =
+  | GetAllSnaps
+  | GetSnap
+  | HandleSnapRequest
+  | HasPermission
+  | GetPermissions;
+
+export type CronjobControllerEvents =
+  | SnapAdded
+  | SnapBlocked
+  | SnapInstalled
+  | SnapRemoved
+  | SnapUnblocked
+  | SnapUpdated
+  | SnapTerminated;
+
+export type CronjobControllerMessenger = RestrictedControllerMessenger<
+  'CronjobController',
+  CronjobControllerActions,
+  CronjobControllerEvents,
+  CronjobControllerActions['type'],
+  CronjobControllerEvents['type']
+>;
+
+const DAILY_TIMEOUT = inMilliseconds(24, Duration.Hour);
+
+export type CronjobControllerArgs = {
+  messenger: CronjobControllerMessenger;
+  /**
+   * Persisted state that will be used for rehydration.
+   */
+  state?: CronjobControllerState;
+};
+
+export type Cronjob = {
+  timer?: Timer;
+  id: string;
+  snapId: SnapId;
+} & CronjobSpecification;
+
+export type StoredJobInformation = {
+  lastRun: number;
+};
+
+export type CronjobControllerState = {
+  jobs: Record<string, StoredJobInformation>;
+};
+
+const controllerName = 'CronjobController';
+
+/**
+ * Use this controller to register and schedule periodically executed jobs
+ * using RPC method hooks.
+ */
+export class CronjobController extends BaseController<
+  typeof controllerName,
+  CronjobControllerState,
+  CronjobControllerMessenger
+> {
+  private _messenger: CronjobControllerMessenger;
+
+  private _timers: Map<string, Timer>;
+
+  // Mapping from jobId to snapId
+  private _snapIds: Map<string, string>;
+
+  private _dailyTimer!: Timer;
```

Comment:
If these need to be private, use `#messenger`, `#timers`, etc. instead of TypeScript's `private` modifier.

Reply:
I made `#messenger` and `#dailyTimer` private in that way. The problem with the others is that I'm using these in the unit tests. 

That is also the reason why there are things like this `expect(cronjobController['_timers'].size).toBe(1);` instead of this `expect(cronjobController._timers.size).toBe(1);`. Otherwise the TypeScript error will appear `error TS2341: Property '_timers' is private and only accessible within class 'CronjobController'.`.

I can see for now that there are two ways for testing this, the one implemented and the other one would be with making these `#` private but also making *getters* which would (at least for now) only be used in tests.

The third way I've already tried is by only calling the entry functions and trying to cover everything. This becomes very difficult because of Jest and asynchronous architectural patterns used with the messenger. Also, it makes test somehow more unreliable since I cannot access the 'storage' variables that are source of truth that are jobs scheduled or removed correctly. It also becomes difficult to make code coverage up to 100% for every branch, so coverage failures would appear. Anyway, all of these hacks are inside the test files, none is in the functional code.

Let me know if there are other ways how you're solving these kind of problems in this repository or if some of my solutions might be acceptable. @Mrtenz 

Also @FrederikBolding you can take a look into this and let me know your opinion.

Reply:
After some discussion `#snapsIds` and `#timers` are made private in that way and getters are added for both.